### ISSUE TITLE
[RFC] Simple notification 'system'

### ISSUE THREAD

**Issue by loevgaard** [Sylius/Sylius#10345]
**Describe the proposed solution**
Basically this screenshot shows everything:

![image](https://user-images.githubusercontent.com/2412177/56964171-ebdf0100-6b5a-11e9-97a9-e7285a5483fd.png)

Is this a good idea you think? Could it be accepted in the core?

Implementation ideas are also very welcome :)

**Comment by Prometee** [comment]
I was wondering how to implement it also but this behaviour is good but if we have other notifications to add it would be better to put this at the top right of the page into a dedicated notification area.

**Comment by Roshyo** [comment]
About the counter itself : 
* Is it a counter of new orders ?
* If so, how would you define if the order has been read or not.
* Will it be related to each admin user or global ?

About the performances : 
* It must be done in AJAX or Websockets to not impact pages performances (that are already way too low in admin, but it's an admin, so it must be ok)

About the notification system : 
* IMO it must be a plugin
* It might be cool as well to provide other notifications, like Email/SMS/Push to Slack ? :)
* Be overridable to allow creating new notifications simply

**Comment by loevgaard** [comment]
The reason I am asking whether to include it in core is because in my opinion it enhances the experience of every aspect of the admin user interface. My worry is that if it's done as a plugin it won't be adopted by the community. As a parallel you can think of the debug toolbar in Symfony. You can write data collectors to show data in the toolbar, but if the toolbar was a separate bundle it wouldn't enjoy the same adoption as it does now. What do the core team think about this, @Zales0123?

Regarding your other points:

> Is it a counter of new orders ?

'Core' notifications would have to be discussed, but I was probably thinking about uncompleted orders, i.e. orders that has to be processed in some way.

> If so, how would you define if the order has been read or not.

Again this has to be discussed, and there is also some implementation detail in this question. But if it was uncompleted orders, an order would have been 'read' when it was completed.

> Will it be related to each admin user or global ?

That is probably also an implementation detail. A scope of some sort. Some notifications would make sense to be user notifications, but others would be global.

> It must be done in AJAX or Websockets to not impact pages performances (that are already way too low in admin, but it's an admin, so it must be ok)

We are on the exact same page regarding performance. It has to be speedy.

> It might be cool as well to provide other notifications, like Email/SMS/Push FTP/Push to Slack ? :)

That would be cool features, yes!

> Be overridable to allow creating new notifications simply

Totally agree!


Thank you very much for your thoughts @Roshyo!

**Comment by loevgaard** [comment]
Regarding your point, @Prometee, there are probably differences between 'textual' notifications as the one in the notification regarding your Sylius version and 'numeric' notifications as how many orders are not processed. And those different notifications would probably have their different places.

**Comment by loevgaard** [comment]
For reference, Laravels notification package is probably interesting to look into: https://laravel.com/docs/5.8/notifications

**Comment by bitbager** [comment]
We have done something like this in the past. The notifications were displayed in the top right corner in the ring bell. Here's how we implemented it:

1. We introduced a trait with `isNew` field that we implemented in Sylius resources we wanted to listen to the new event. 
2. We created a Request event listener that was executed each time we opened a website. As the admin panel is based on CRUD operations in ResourceController, we created a nice architecture with injected model classes parameters (i.e. `%sylius.order.model.class%`) that are handled by this listener.
3. We created a generic repository that allows injecting resource repositories in it and builds a query upon the `isNew` field.
4. We created a notification controller that glues everything together and rendered it in the corresponding admin notification layout.
5. :tada:

**Comment by 4c0n** [comment]
I think it would great to have this feature and yes I also think it should be part of Sylius itself.

As a basic first version it would be awesome to see that this just works on page loads (if the page load time doesn't increase significantly, so I guess that would have to be benchmarked and compared to some baseline tests). Without knowing what the difference in execution time is in different situations, I don't think we can state that this *has* to be done with async calls for performance reasons (which could potentially also be overcome by using caching for example).

Choosing for an asynchronous load of that information would require adding some form of communication that is currently not necessarily implemented (like Mercure, Pusher, WebSockets, ...), which would be my preferred way of keeping this info up to date without having to reload the page, so I guess using an AJAX call, would be the only real option when this should be ending up in the core, IMO.

Unless of course the Sylius team is willing to add a dependency on another system/protocol, but I doubt that. In which case it would have to become a plugin so that users can choose.

All in all it would probably be wise to design a proper notification system in Sylius and perhaps build it in such a way that it can be used with different techniques, like page loads, AJAX, Event-Stream, WebSocket, etc... Allowing users to choose how to implement their notifications in their frontend. (So not just admin, but also allowing this to work in the user side of the shop).

**Comment by loevgaard** [comment]
Very good comments, @4c0n :)

**Comment by gabiudrescu** [comment]
hi guys,

I would vote against this being part of the core - at least for now. let me explain my decision:
 - imho, the Sylius Core team should focus on something else right now: implement the features on the current roadmap, create the skeleton theme of Sylius that could help new comers to better design their UI, migrate from SemanticUI to a newer and better supported UI framework (eg: Bootstrap maybe?) etc.
 - having badge notifications in menu is a big assumption to make - at least, from Sylius Core point of view; for orders: you assume that people would be interested in seeing there a new badge whenever a new order appears that require attention from the admin operator; but not all admin operators are the same; and implementing such a custom logic, brings me to the first point (this should not be in the Core Team responsibility)

having a plugin that allows me to implement badge notifications based on various scenarios seems like a better idea to me.

**Comment by 4c0n** [comment]
@gabiudrescu Not to attack you or anything, but:
"imho, the Sylius Core team should focus on something else right now", you could always say that. There is no timeline mentioned in this issue, so I'm sorry, but it might not be that relevant to use as a valid argument in this discussion. Also it does not necessarily have to be the Sylius team that implements this. Please correct me if I'm wrong.

"having badge notifications in menu is a big assumption to make - at least, from Sylius Core point of view; for orders: you assume that people would be interested in seeing there a new badge whenever a new order appears that require attention from the admin operator; but not all admin operators are the same; and implementing such a custom logic, brings me to the first point (this should not be in the Core Team responsibility)"

I totally see your point here. I agree that not all shop admins are the same. However, if the idea is to build a generic notification backend (multi purpose), it is also possible to make something like this configurable and opt-out of having these notfications for example. Even for making these type of plugins possible/easy to develop, there should be some way of dealing with notifications. Sylius providing a way to do that would stimulate people to actually build plugins like that.
I do think having a plugin just to show this one little notfication is a bit overkill though.

**Comment by loevgaard** [comment]
@gabiudrescu regarding your comments:

> imho, the Sylius Core team should focus on something else right now:

On this matter I agree with @4c0n. If an idea is *good* and people agree it should be in the core, then it should. The timeline is another matter.

> having badge notifications in menu is a big assumption to make

I guess everybody sees your point, however, I don't think it's a valid argument _against_ this feature. I think it more describes a concern that should be thought through before implementing the feature.

Say you have an admin who is a 'product creator'. In the current admin there are **a lot** of menu items that he/she should ignore (because he basically only needs the `Products` view). So adding a badge to orders (which is already ignored by him) shouldn't be an argument against this feature. With the RBAC plugin I guess you could also limit what he sees anyway (I haven't tried the RBAC plugin)?

Also I believe that the functionality could/should be implemented in a way where notifications could be disabled or extended.

**Comment by loevgaard** [comment]
Very good discussion around this topic. Thank you guys.

For me it looks like it boils down to: Should Sylius have a native (core) notification system?

In my opinion it should. My arguments are:

1. The user experience would benefit from it. I believe in 2019 people are used to notifications everywhere and for me at least it enhances user interfaces when 'real time' updates prompts me to take action or shows that I should take action. Just to give you a peek inside my head regarding use cases here are a few:
   - When installing a plugin that needs setup you can show a badge in the menu that indicates that you should take action to make it work.
   - A badge could show the number of uncompleted orders
   - If you have cron jobs running or jobs in a message queue you could use the notification to tell the user that that job has finished and give a link to the results (if any)

2. If you agree that the overall user experience would benefit from it then you'd also want the adoption to be as high as possible and that would be much more likely if it was implemented in the core.

**Comment by Roshyo** [comment]
> imho, the Sylius Core team should focus on something else right now:

Indeed, that's not a valid point (even though it is true). And still, even the fact that it isn't in the core shouldn't mean CoreTeam won't work on it. RBAC and Invoicing plugins are examples of that.

However, as Sylius stated many times they focus on core e-commerce functionalities. And this is definitely not a core one.

I'd agree if it was already implemented in a very basic way of badges displaying an amount. But as we saw it, it can (and should) go way more further than just a badge. Indeed notifications are definitely a huge (mandatory) improvement in any system. It also covers a lot of functionalities and technologies (AJAX, Websocket, SMS, Mail, API, ...) and then should not be Sylius Core's concern but rather an highly customizable plugin one.


> When installing a plugin that needs setup you can show a badge in the menu that indicates that you should take action to make it work.

I believe this belongs to another use case of "PluginManagement" that should indeed rely on Notification System. But plugin dependency isn't forbidden nor hard to do in Sylius.


Your last point is clearly valid, but I think that a choice like this shouldn't be made according to this reason.

**Comment by gabiudrescu** [comment]
Before 1.0 was launched, Sylius got a lots of useful features trimmed and removed from core, just to arise at a later points as standalone plugins; eg: Reporting, RBAC. 

This doesn't mean they are less important or not useful. They just don't belong to core, as they might have different release cycles than core. we don't want a full core release delayed by a feature that is not mandatory for Sylius to work. 

Although initially I wasn't very happy to see Reporting and RBAC removed from Core, with time, I ended up understanding the rationale behind this decision. And I've become more and more a supporter of the modularity the plugin system allows. 

@4c0n no offence taken, it's just a civilized discussion between people with different opinions. no worries. 

now, let's take them one by one:

>  Also it does not necessarily have to be the Sylius team that implements this
It doesn't matter who implements it. as long as it is in the core, it will be the Core Team  responsibility in the future to maintain it. 

And given their size and focus right now, I don't think extra features that are not mandatory should be on their responsibility. 

Having it implemented as a plugin backed by the Core Team - well, that's a totally different ball game. 

> I do think having a plugin just to show this one little notification is a bit overkill though.

we both agree on this one. my proposal was to have a plugin for an abstract notifications system that can be used to implement simple or more complex logic for notifications. 

> In the current admin there are a lot of menu items that he/she should ignore (because he basically only needs the Products view)

use RBAC for this to reduce the clutter in the menu rather than add some notifications to catch the attention of someone who is scrolling 4 screens to reach what they need to reach. 

we use RBAC to reduce the number of menu items with great success.

> you'd also want the adoption to be as high as possible and that would be much more likely if it was implemented in the core.

in my opinion, it should be the other way around: if it's popular enough, include it in the core. because then you know it will really contribute to the value of the core and you want to give it more support.

struggling to make something popular by including it to core feels wrong to me. 

----

Take away notes:

- I'm not against the idea of this functionality; I think it brings value to the project; on a scale of value from 1 to 10, where 1 is some really obscure feature nobody really is using and 5 is basic ecommerce stuff like orders and products, I would give this feature an honorable 2. my personal take, don't throw stones
 - I strongly think this should be a standalone plugin and, after it proved it's usefulness to the entire community, propose to have it included in the core
 - the Core Team effort is limited by their capacity (number of people, billable hours they can spend on various functionalities); based on various conversations I had with some of the Core Team members (eg: Product Owner, marketing, Pawel), my feeling is that their attention should be now on something else rather than such new features that are good to have, but not mandatory for success
 - put yourself in the shoes of a business owner: how much would you be willing to pay for such a feature? let's say 5k € (that's 100h of development on a 50€/h rate); isn't there anything else you would do with those money in Core in order to make Sylius better for the *entire* community? just to throw some ideas:
 -- stable full shop API
 -- return merchandise
 -- import / export module for PIM, ERP and other 3rd party integrations
 -- better default theme
 -- marketplace module
 -- data audit
 -- WYSIWYG content editor for the website

spending 100h of development on any of those ideas in core is, in my opinion, better for the entire community rather than implementing a new (and useful feature) such as notifications. 

spending 100h of development on notifications as a plugin seems to me like a better idea. 

I think we should move this discussion forward to how the notification system should behave rather than argue infinitely on where this should be implemented.

**Comment by vvasiloi** [comment]
It's a surprisingly intensive discussion for that topic, so I would like to add my two cents.
Since I'm quite active on the Slack server, especially in the #support channel, trying to help other members of the community, I can say that there are enough issues/features to work on beside this one.

People are struggling customising and extending Sylius and sometimes even struggling to make it work out-of-the-box.
Most of the time it's about the lack of details in the documentation, sometimes it's non-existent, especially when it comes to technical details.

There were a few cases when people were strugling to use standalone components and bundles, or use just the "core" for APIs and other scenarios when the admin and/or shop is not needed.
Besides the same lack of documentation, another bottleneck is still some lack of decoupling here and there.

By adding more features to core, there is potentially more coupling between the components, because mistakes can be made and that's normal.
It would be nice to have some notifications in Sylius, but I think it's better to start it as plugin.

I like Sylius as it is, with it's obviosly existing "space to improve", mostly due to it's DX and because it has "No “Not Invented Here” Syndrome" - @astorm (https://alanstorm.com/five-first-impressions-of-the-sylius-ecommerce-system/)
That being said, I think that plugin should aim to integrate existing solutions into Sylius, rather than inventing it's own thing.
There are a lot of libraries that can be reused for the the notification system itself and even more for integrate it with different notification channels.
To name a few:
https://github.com/irishdan/NotificationBundle
https://github.com/sonata-project/SonataNotificationBundle
https://github.com/Seldaek/monolog (yeah, even that one)

**Comment by gabiudrescu** [comment]
@vvasiloi I think @loevgaard was talking about badge notifications rather than flash notifications. eg:
this 
![image](https://user-images.githubusercontent.com/5156054/57317274-6205d980-7100-11e9-8534-9862cedf42f9.png)

instead of
![image](https://user-images.githubusercontent.com/5156054/57317246-574b4480-7100-11e9-99af-da4f6486bb66.png)

though I must say, I tend to agree with what you say.

**Comment by vvasiloi** [comment]
@gabiudrescu I know he was talking only about the badges, but there were suggestions of making a more generic/abstract notification. It could potentially cover both badges and flashes (which can be considered a channel).

My "two cents" can be resumed to:
- not a priority nor responsability of the core team right now
- start as a plugin
- do not reinvent the wheel
- aim to "glue" existing solutions into Sylius

I mentioned [monolog ](https://github.com/Seldaek/monolog) because that's probably the most common [ PSR-3](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md) implementation, which I think will be a good "contract" for the notification plugin and it has a lot of [integrations ](https://github.com/Seldaek/monolog/blob/master/doc/02-handlers-formatters-processors.md).

The [SonataNotificationBundle ](https://github.com/sonata-project/SonataNotificationBundle) is nice, but it's designed for usage with queues and other Sonata bundles.

https://github.com/irishdan/NotificationBundle looks good at the first sight, but I have to review it in depth to give a valuable feedback.

Any of those can be used at least as a source of inspiration.

Those who agree - start throwing ideas, requiremets, libraries, etc. :)

**Comment by 4c0n** [comment]
Giving it some more thought, making a notification plugin would make all
sorts of sense. Given decoupling, etc... it would become a "bundle" anyway
because it would be useful on its own.

As a rough sketch:

- use events to publish notifications
- allow different transportation methods (queues, Mercure, custom, whatever)
- allow configuration of notifications
- allow for different types of notifications
- more??

Something else to consider might be if and how the frontend side of this
should be included.

**Comment by stefandoorn** [comment]
After just reading the whole conversation (which is rather interesting to read), I'll also share my two cents. 

* I think (as a lot of 'features') this should be a separate plugin. Later on it could be distributed via Sylius-Standard if that makes sense (as @gabiudrescu mentioned). Core should be such extensible that this plugin can be implemented, which makes it automatically suitable for other extensions also. Core benefits from such plugins, but the plugin itself should be stand-alone initially.
* I think we should differentiate wording. I think @loevgaard started the discussion with 'badges', not really notifications. It's wording, but it's good to limit plugin functionality - especially in the beginning - in order to make it to a deliverable. Feels more 'agile' to me.
* I totally agree that such feature should not slow down the admin environment, but I only see two proper options to start: either in the pageview directly, or via AJAX calls. I don't like that much AJAX calls, as user can still see all kind of things moving around after the initial page load. Also, how many requests are that going to be, what happens if the connection is slow, how much extra connections do we open, etc.? I think this kind of notifications should be fast enough to load on page render. There could be some caching, maybe even invalidate cache on certain Sylius events. Not sure what's the best method here, but I think going with the initial pageview feels more 'safe', easier to implement, easier to reach a deliverable, can still be tested (and such way later the implementation can safely change), and new transport methods can be configurable. 
* I think it should start with simple counters (e.g. unfinished orders) without admin user state. Also, which badges are visible should be configurable with a common default configuration. Next to this, the 'fetchers' of the badge count should be easy to decorate, in order to change behavior smoothly. E.g. in one of the shops I work on, we don't report on 'completed' orders, but on an extra status that we've implemented in between. E.g. our counter makes more sense to show count of orders that have not reached that status yet. I don't mind that the plugin doesn't do that out of the box, but I wouldn't like it that it takes lot customisations.

Anyway, good feature I think, as long as we keep it simple in the beginning with proper extension points. 

P.S. #8406 got upvoted today again, which might be handy to have in case anyone wants to implement  it via streams / AJAX btw. The more extension points, the easier plugins are to implement.

**Comment by alexrowesoap** [comment]
Another viewpoint from myself: I manage a very large WC store that has a messaging plugin. By far and a way the biggest issue is the query for getting the new message count. Partially this is down to WP's DB schema and having to calculate this on the fly, but it's still an overhead. As an admin, you get a notification on the number of unread messages, and with the number of messages present, it's very intensive and slow if calculated at run-time. 

If I could turn back time I would have made an emitter / listener system (to replace the vendors)  wherein the role of the user defines the listeners associated but always accepts direct notifications. This would then ideally be stored in a dedicated table, consisting of the recipient user, the count and something similar to a class name or similar to associate to.

TLDR: Don't calculate notifications on objects at runtime: store them somewhere to both speed up lookups and decouple the target of the notification.

I'd also like it to be a bundle for now, just to keep core complexity to a minimum. Don't get me wrong, it's a sweet idea, but in _my_ use-cases, it's a nice-to-have as opposed to required functionality IMO.

**Comment by alexrowesoap** [comment]
Thanks @Zales0123 :)
